### PR TITLE
chore: bump typescript-eslint to v8

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
👋 Hi! Coming over from https://github.com/typescript-eslint/typescript-eslint/issues/9501: we're working on typescript-eslint v8 and would like to try the beta out on plugins such as `eslint-plugin-jest`.

I'm sending this draft PR as a reference to see what issues we might give you in upgrading and to try to be helpful. If this is annoying noise to you, please forgive me, I'll withdraw the PR. But I hope this is useful on your end - and please let me know if you have any feedback! ❤️

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -30,7 +30,9 @@ module.exports = {
     '@typescript-eslint/array-type': ['error', { default: 'array-simple' }],
     '@typescript-eslint/no-require-imports': 'error',
     '@typescript-eslint/ban-ts-comment': 'warn',
-    '@typescript-eslint/ban-types': 'error',
+    '@typescript-eslint/no-empty-object-type': 'error',
+    '@typescript-eslint/no-unsafe-function-type': 'error',
+    '@typescript-eslint/no-wrapper-object-types': 'error',
```
Comment (JoshuaKGoldberg) [inline .eslintrc.js:35]: `ban-types` is removed in typescript-eslint v8. Its default options are now split among those three rules.

File: src/rules/utils/misc.ts [hunk 2]
```diff
@@ -16,7 +16,11 @@ import { type ParsedExpectFnCall, isTypeOfJestFnCall } from './parseJestFnCall';
 
 const REPO_URL = 'https://github.com/jest-community/eslint-plugin-jest';
 
-export const createRule = ESLintUtils.RuleCreator(name => {
+export interface JestPluginDocs {
+  requiresTypeChecking?: boolean;
+}
+
+export const createRule = ESLintUtils.RuleCreator<JestPluginDocs>(name => {
```
Comment (JoshuaKGoldberg) [inline src/rules/utils/misc.ts:23]: https://typescript-eslint.io/blog/announcing-typescript-eslint-v8-beta/#custom-rule-metadocs-types

In theory there's no hard need in plugins to mention `requiresTypeChecking` anymore... except, in practice `eslint-doc-generator` reads it:

https://github.com/jest-community/eslint-plugin-jest/blob/4cc2a1b680507ff006b5d2b02fa6d262584bb263/.eslint-doc-generatorrc.js#L25

**Comment by SimenB** [comment]
>If this is annoying noise to you, please forgive me, I'll withdraw the PR.

Not at all! Please keep the engagement up 🙂

**Comment by notaphplover** [comment]
@JoshuaKGoldberg are you working on this? typescript-eslint 8 is released and it would be great to update this repo to add support for it. If you are too busy to go for it, would it be ok to submit another PR to add support for typescript-eslint 8?

File: package.json [hunk 3]
```diff
@@ -106,7 +106,7 @@
     "typescript": "^5.0.4"
   },
   "peerDependencies": {
-    "@typescript-eslint/eslint-plugin": "^6.0.0 || ^7.0.0",
+    "@typescript-eslint/eslint-plugin": "*",
```
Comment (notaphplover) [inline package.json:109]: Are we sure every `@typescript-eslint/eslint-plugin` would be a valid peer?

**Pull request closed without merging**